In [1]:
from google.colab import auth
auth.authenticate_user()

import gspread
from google.auth import default
creds, _ = default()

gc = gspread.authorize(creds)

資料來源https://docs.google.com/spreadsheets/d/103Q1qgG_RKBX3Bf4jWmWO8FZP6wePTu33eRLMUwOUlI/edit?usp=sharing

In [2]:
import pandas as pd

gsheets = gc.open_by_url('https://docs.google.com/spreadsheets/d/103Q1qgG_RKBX3Bf4jWmWO8FZP6wePTu33eRLMUwOUlI/edit?usp=sharing')

In [5]:
gsheets

<Spreadsheet '110學年(下) 高三學測成績' id:103Q1qgG_RKBX3Bf4jWmWO8FZP6wePTu33eRLMUwOUlI>

In [7]:
rows = worksheet.get_all_values()
for i, row in enumerate(rows[:5]):
    print(f"第 {i+1} 列：{row}")

第 1 列：['編號', '學號', '學校班級', '姓名', '錄取學校 ', '', '', '', '國文', '英文', '數A', '數B', '社會', '自然', '英聽', '入取管道']
第 2 列：['1', '1500013', '新莊302', '許家瑋', '國立臺灣師範大學 歷史學系', '', '', '', '11(均+1)', '12(前)', '7(均+1)', '12(前+1)', '13(頂)', '9(後+2)', 'B', '分科']
第 3 列：['2', '1500022', '海山309', '李冠緯', '國立臺灣師範大學 英文學系', '', '', '', '13(頂)', '14(頂+1)', '7(均+1)', '12(前+1)', '13(頂)', '6(底+1)', 'A', '分科']
第 4 列：['3', '1500026', '林口303', '楊席紘', '大同大學 事業經營學系', '', '', '', '10(均)', '9(均+1)', 'X', '3(底)', '10(後+1)', 'X', 'C', '申請']
第 5 列：['4', '1500027', '北一二儉', '吳怡晨', '國立中興大學 獸醫學系', '', '', '', '14(頂+1)', '14(頂+1)', '8(前)', 'X', '11(均)', '13(前+1)', 'A', '分科']


In [8]:
# 開啟 Google Sheet
gsheets = gc.open_by_url('https://docs.google.com/spreadsheets/d/103Q1qgG_RKBX3Bf4jWmWO8FZP6wePTu33eRLMUwOUlI/edit?usp=sharing')
worksheet = gsheets.get_worksheet(0)

# 取得所有資料
rows = worksheet.get_all_values()

# 自訂欄位名稱（略過空欄）
headers = ['編號', '學號', '學校班級', '姓名', '錄取學校', '國文', '英文', '數A', '數B', '社會', '自然', '英聽', '入取管道']

# 因為前面有 3 個空欄，所以要取 index 的方式來對應正確欄位
data_rows = []
for row in rows[1:]:  # 從第 2 列開始是資料
    # 組合正確欄位（跳過 index 5~7 的空值）
    filtered_row = row[0:5] + row[8:]
    data_rows.append(filtered_row)

# 轉為 DataFrame
df = pd.DataFrame(data_rows, columns=headers)


In [9]:
df

編號       學號   學校班級   姓名           錄取學校       國文       英文      數A       數B  \
0    1  1500013  新莊302  許家瑋  國立臺灣師範大學 歷史學系  11(均+1)    12(前)  7(均+1)  12(前+1)   
1    2  1500022  海山309  李冠緯  國立臺灣師範大學 英文學系    13(頂)  14(頂+1)  7(均+1)  12(前+1)   
2    3  1500026  林口303  楊席紘    大同大學 事業經營學系    10(均)   9(均+1)       X     3(底)   
3    4  1500027   北一二儉  吳怡晨    國立中興大學 獸醫學系  14(頂+1)  14(頂+1)    8(前)        X   
4    5  1500029  中山302  陳沛誼     國立成功大學 化學系    12(前)    12(前)  5(後+1)  14(頂+1)   
..  ..      ...    ...  ...            ...      ...      ...     ...      ...   
65  66  1500136  新莊301  黃琳雅      輔仁大學 法律學系                                      
66  67  1500138   華興三甲  王芊昀    中原大學 特殊教育學系                                      
67  69  1500137  丹鳳306  王睿賢    中原大學 土木工程學系  11(均+1)     5(後)    4(後)   6(後+2)   
68  70  1500139  新莊317  張昭淳    台東大學 幼兒教育學系  11(均+1)  10(均+2)    8(前)  12(前+1)   
69  71                                                                          

         社會       自然 英聽 入取管道  
0     13(頂)   9(後+2)  B   分科  
1     13(頂)   6(底+1)  A   分科  
2   10(後+1)        X  C   申請  
3     11(均)  13(前+1)  A   分科  
4     13(頂)  11(均+1)  B   分科  
..      ...      ... ..  ...  
65                    A   申請  
66                        申請  
67    12(前)   9(後+2)  C   申請  
68        X    12(前)  B   申請  
69                            

[70 rows x 13 columns]

In [10]:
# 欄位標題（略過空白欄）
headers = ['編號', '學號', '學校班級', '姓名', '錄取學校', '國文', '英文', '數A', '數B', '社會', '自然', '英聽', '入取管道']

# 過濾空欄（index 5~7）後取出資料
data_rows = []
for row in rows[1:]:
    filtered_row = row[0:5] + row[8:]
    data_rows.append(filtered_row)

# 建立 DataFrame
df = pd.DataFrame(data_rows, columns=headers)

# 要清洗的成績欄位
score_columns = ['國文', '英文', '數A', '數B', '社會', '自然']

# 移除括號與內文，只保留數字
for col in score_columns:
    df[col] = df[col].str.extract(r'(\d+)')  # 抓出數字部分
    df[col] = pd.to_numeric(df[col], errors='coerce')  # 轉成數值型態，非數字轉 NaN

# 顯示前幾筆清洗後的資料
df.head()

編號       學號   學校班級   姓名           錄取學校    國文    英文   數A    數B    社會    自然  \
0  1  1500013  新莊302  許家瑋  國立臺灣師範大學 歷史學系  11.0  12.0  7.0  12.0  13.0   9.0   
1  2  1500022  海山309  李冠緯  國立臺灣師範大學 英文學系  13.0  14.0  7.0  12.0  13.0   6.0   
2  3  1500026  林口303  楊席紘    大同大學 事業經營學系  10.0   9.0  NaN   3.0  10.0   NaN   
3  4  1500027   北一二儉  吳怡晨    國立中興大學 獸醫學系  14.0  14.0  8.0   NaN  11.0  13.0   
4  5  1500029  中山302  陳沛誼     國立成功大學 化學系  12.0  12.0  5.0  14.0  13.0  11.0   

  英聽 入取管道  
0  B   分科  
1  A   分科  
2  C   申請  
3  A   分科  
4  B   分科

In [11]:
pip install scikit-learn plotly


In [12]:
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
import plotly.express as px

In [13]:
# 做 KMeans 聚類（假設你想分成 3 群）
kmeans_model = KMeans(n_clusters=3, random_state=42, n_init=10)
cluster_features = df[score_columns].dropna()
df_valid = df.loc[cluster_features.index]
df_valid['cluster'] = kmeans_model.fit_predict(cluster_features)

# PCA 降維
pca = PCA(n_components=2)
pca_result = pca.fit_transform(cluster_features)
df_valid['pca1'] = pca_result[:, 0]
df_valid['pca2'] = pca_result[:, 1]

# 畫圖
fig = px.scatter(
    df_valid, x='pca1', y='pca2',
    color=df_valid['cluster'].astype(str),
    hover_data=['姓名', '學校班級', '錄取學校'],
    title='學生成績 KMeans 聚類 + PCA 可視化'
)
fig.show()

# 學生成績 KMeans 聚類 + PCA 可視化（含各科成績）

🔵 藍色	0	成績分布較平均或居中，主成分值落在右側，有可能是「表現優異」的群體。例如:海山309  

🟢 綠色	1	多數集中在左下方，PCA 值偏負，可能代表「成績整體較低」的群體。如三重三律

🔴 紅色	2	分布在中間偏右，有可能代表「某幾科特別強或偏科」的學生群。如新莊316








各群趨勢細節

✅ 紅色群組（color=2）：

明顯集中在圖的右中區。

代表這些學生的 PCA 主成分分數較高，某些成績特別突出。

### 看起來這群學生的成績變異度較大，可能是某些科目特別強（如自然或數學A高分）。





✅ 綠色群組（color=1）：

集中在圖的左側偏下。pca1 和 pca2 多為負值，表示整體成績較低或在主成分上無優勢。

### 這群學生可能是整體學科表現偏弱的群體。





✅ 藍色群組（color=0）：

分散程度較大，但多數集中在右下方。主成分值整體偏正，成績可能 均衡且中上程度。

### 這群學生可能是「穩定發揮」的類型，各科不偏科。

In [19]:
fig = px.scatter(
    df_valid,
    x='pca1', y='pca2',
    color=df_valid['cluster'].astype(str),
    hover_data=[
        '學校班級', '錄取學校',
        '國文', '英文', '數A', '數B', '社會', '自然'  # 加入科目成績
    ],
    title='學生成績 KMeans 聚類 + PCA 可視化（含各科成績）'
)
fig.show()


 圖表解讀：以「國文」與「英文」為基礎的 KMeans 分群散點圖

 軸向與變數：

X 軸：國文成績

Y 軸：英文成績

顏色：KMeans 分群（0：藍、1：綠、2：紅）


群 0（藍）	國文成績高，英文成績也高

群 1（綠）	國文成績中等，英文成績較低

群 2（紅）	國文成績中～高，英文成績中等

🧠 各群組的趨勢分析：

🔵 群 0（藍色）：
分布在 國文 12到14 級，英文 11到14 級 的區域。

特徵：國文與英文都表現非常好。

**這群學生屬於「雙語強者」，語文能力明顯突出，是整體表現最佳的一群。**

🟢 群 1（綠色）：
分布集中在 國文 8到11 級，英文 4到7 級。

特徵：國文中等，英文偏低。

**這群學生可能偏好中文學習，英文為相對弱項，屬於「中語文、弱英文型」。**

🔴 群 2（紅色）：
分布在 國文 9到13 級，英文 5到10 級。

特徵：國文略高於英文，但兩者差距不如群 1 明顯。

**屬於「語文平均型」，語文整體普通，未特別強弱。**



In [18]:
import plotly.express as px

fig = px.scatter(
    df_valid,
    x='國文',
    y='英文',
    color=df_valid['cluster'].astype(str),
    hover_data=['姓名', '學校班級', '錄取學校'] + score_columns,
    title='KMeans 分群結果：以原始科目成績為散點圖'
)
fig.show()

Grouping and Counting Behaviors per Theme